packages/preferences/src/browser/folders-preferences-provider.ts

```diff
@@ -0,0 +1,140 @@
+/********************************************************************************
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, postConstruct } from 'inversify';
+import { PreferenceProvider } from '@theia/core/lib/browser';
+import { ILogger } from '@theia/core/lib/common';
+import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service';
+import { FolderPreferenceProvider, FolderPreferenceProviderFactory } from './folder-preference-provider';
+import { FileSystem, FileStat } from '@theia/filesystem/lib/common';
+import URI from '@theia/core/lib/common/uri';
+
+@injectable()
+export class FoldersPreferencesProvider extends PreferenceProvider {
+
+    @inject(ILogger) protected readonly logger: ILogger;
+    @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService;
+    @inject(FileSystem) protected readonly fileSystem: FileSystem;
+    @inject(FolderPreferenceProviderFactory) protected readonly folderPreferenceProviderFactory: FolderPreferenceProviderFactory;
+
+    private providers: FolderPreferenceProvider[] = [];
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        await this.workspaceService.roots;
+        if (this.workspaceService.saved) {
+            for (const root of this.workspaceService.tryGetRoots()) {
+                if (await this.fileSystem.exists(root.uri)) {
+                    const provider = await this.createFolderPreferenceProvider(root);
+                    this.providers.push(provider);
+                }
+            }
+        }
+
+        // Try to read the initial content of the preferences.  The provider
+        // becomes ready even if we fail reading the preferences, so we don't
+        // hang the preference service.
+        Promise.all(this.providers.map(p => p.ready))
+            .then(() => this._ready.resolve())
+            .catch(() => this._ready.resolve());
+
```

Comment:
it would be the best if we could do 
```ts
Promise.all(...blah...)
  .finally(() => this._ready.resolve());
```
but `Promise.finally` is unavailable in this project for now.

The reason that we want to do `ready.resolve()` in both cases is that, we want to unblock the clients from consume the data as long as all providers finish loading. The loading being successful or not successful is not a concern here, as we don't want any features in Theia to hang up if a preference value is (temporarily) inaccessible.